### PR TITLE
Edit Feature Attachment - Sample

### PR DESCRIPTION
**Pull request by manishk3189**
Hi @doneill, I have created a sample to demonstrate how to add, delete or fetch attachments for a feature in reference to [#991](https://devtopia.esri.com/runtime/java-android-sdk/issues/991). Can you please review the same and let me know of changes if required?

Thanks

### CODE REVIEW

File: edit-feature-attachments/build.gradle [hunk 1]
```diff
@@ -0,0 +1,27 @@
+apply plugin: 'com.android.application'
+
+android {
+    compileSdkVersion rootProject.ext.compileSdkVersion
+    buildToolsVersion rootProject.ext.buildToolsVersion
+
+    defaultConfig {
+        applicationId "com.example.editfeatureattachments"
+        minSdkVersion 16
+        targetSdkVersion 22
```
Comment (doneill) [inline edit-feature-attachments/build.gradle]: Not targeting latest version of Android, test against latest and update.

File: edit-feature-attachments/build.gradle [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+        }
+    }
+}
+
+dependencies {
+    compile fileTree(dir: 'libs', include: ['*.jar'])
+    testCompile 'junit:junit:4.12'
+    compile 'com.android.support:appcompat-v7:23.4.0'
+    compile 'com.android.support:design:23.4.0'
```
Comment (doneill) [inline edit-feature-attachments/build.gradle]: Newer version of support libs available, test and update these.

File: edit-feature-attachments/src/main/res/layout/attachment_entry.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        android:layout_marginLeft="5dp"
+        android:layout_marginRight="5dp"
+        android:layout_marginTop="5dp"
+        android:src="@mipmap/map_place_holder" />
+    <TextView
+        android:id="@+id/AttachmentName"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:paddingTop="5dp"
+        android:singleLine="true" />
```
Comment (doneill) [inline edit-feature-attachments/src/main/res/layout/attachment_entry.xml]: `singleLine` deprecated, use `maxLines="1"` instead.

File: edit-feature-attachments/src/main/res/layout/attachment_entry.xml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+</LinearLayout>
```
Comment (doneill) [inline edit-feature-attachments/src/main/res/layout/attachment_entry.xml]: This layout can be replaced with a single `<TextView/>` and compound drawable.  

```xml
<TextView
     android:id="@+id/AttachmentName"
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
     android:paddingTop="5dp"
     android:maxLines="1"
     android:drawableLeft="@mipmap/map_place_holder"
     android:padding="5dp" />
```

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        this.context = context;
+        this.attachmentName = attachmentList;
+
+
+
+    }
+    @Override
+    public View getView(int position, View view, ViewGroup parent) {
+        LayoutInflater inflater = context.getLayoutInflater();
+        View rowView= inflater.inflate(R.layout.attachment_entry, null, true);
```
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java]: Refactor to use [View Holder pattern](https://developer.android.com/training/improving-layouts/smooth-scrolling.html) for smoother scrolling.

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.esri.arcgisruntime.sample.editfeatureattachments;
```
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java]: Package can be local.

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+import java.util.List;
+
+public class MainActivity extends AppCompatActivity {
+
+    private static final String TAG = "EditFeatureAttachment";
+    private static final int REQUEST_CODE = 100;
+    ServiceFeatureTable mServiceFeatureTable;
+    ProgressDialog progressDialog;
+    RelativeLayout mCalloutLayout;
+    ArcGISMap mMap;
```
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/MainActivity.java]: 4 fields above can be private.

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+public class EditAttachmentActivity extends AppCompatActivity {
+
+    private static String TAG = "EditAttachmentActivity";
+    private static int RESULT_LOAD_IMAGE = 1;
```
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java]: Make above 2 fields final.

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        super.onActivityResult(requestCode, resultCode, data);
+
+        if (requestCode == RESULT_LOAD_IMAGE && resultCode == RESULT_OK && null != data) {
+            Uri selectedImage = data.getData();
+            String[] filePathColumn = {MediaStore.Images.Media.DATA};
+
+            Cursor cursor = getContentResolver().query(selectedImage,
+                    filePathColumn, null, null, null);
+            cursor.moveToFirst();
```
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java]: You need to check that `cursor` is not null

```java
if (cursor != null && cursor.moveToFirst()) {
        // proceed
    }

```

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+                            String fileName = attachmentList.get(position);
+                            // create a drawable from InputStream
+                            Drawable d = Drawable.createFromStream(listenableFuture.get(), fileName);
+                            // create a bitmap from drawable
+                            Bitmap bitmap = ((BitmapDrawable) d).getBitmap();
+                            File root = null;
+                            root = Environment.getExternalStorageDirectory();
+                            File fileDir = new File(root.getAbsolutePath() + "/ArcGIS/Attachments");
+                            // create folder /ArcGIS/Attachments in external storage
+                            fileDir.mkdirs();
```
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java]: You need to test the result of this before moving on.

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+                listenableFuture.addDoneListener(new Runnable() {
+                    @Override
+                    public void run() {
+                        try {
+                            String fileName = attachmentList.get(position);
+                            // create a drawable from InputStream
+                            Drawable d = Drawable.createFromStream(listenableFuture.get(), fileName);
+                            // create a bitmap from drawable
+                            Bitmap bitmap = ((BitmapDrawable) d).getBitmap();
+                            File root = null;
```
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/EditAttachmentActivity.java]: Don't need to initialize as null here.

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java (hunk 5 above)
Comment (manishk3189) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java]: @doneill Should the sample use Cards to display the attachments or ListView is fine?

**Comment by doneill** [comment]
@manishk3189 I have finished my first pass at reviewing this sample.  In addition to the comments above, a couple of things:  

- The drawables are not in density folders
- The edits are not persisting, confirm if this is an issue with the service and detail here.

**Comment by manishk3189** [comment]
@doneill I will refactor the PR accommodating the above comments. 
I also checked the damage assessment service for editing the attachment, and I am getting **unable to complete** error message. It seems like an issue with the service. 
Also, I will change the viewpoint to the US instead of Europe so that features are visible when the app is launched.

**Comment by doneill** [comment]
Thanks @manishk3189 I forgot to mention changing the viewpoint.  I will be looking into this service, in the mean time refactor as appropriate and let me know when ready for another review.

File: edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java (hunk 5 above)
Comment (doneill) [inline edit-feature-attachments/src/main/java/com/esri/arcgisruntime/sample/editfeatureattachments/CustomList.java]: @manishk3189 ListView is fine, more interested in using the View Holder pattern here.

**Comment by manishk3189** [comment]
@doneill I have refactored the sample as per your review comments. 
However, I didn't refactor the compound drawable because the imageView is not just a placeholder but downloaded attachments are set as a bitmap to the image view. Refactoring the XML to compound drawable format and setting the downloaded image distorts the attachment list as only downloaded attachments will have the image not not rest of them.

**Comment by doneill** [comment]
@manishk3189 Looks good will merge.

**Pull request merged**